### PR TITLE
Use less syscalls in `anon_pipe()`

### PR DESCRIPTION
**Pull request by tbu-**
Save a `ENOSYS` failure from `pipe2` and don't try again.

Use `cvt` instead of `cvt_r` for `pipe2` - `EINTR` is not an error
`pipe2` can return.

### CODE REVIEW

File: src/libstd/sys/unix/pipe.rs [hunk 1]
```diff
@@ -29,34 +30,33 @@ pub fn anon_pipe() -> io::Result<(AnonPipe, AnonPipe)> {
     // CLOEXEC flag is to use the `pipe2` syscall on Linux. This was added in
     // 2.6.27, however, and because we support 2.6.18 we must detect this
     // support dynamically.
-    if cfg!(target_os = "linux") {
+    static TRY_PIPE2: AtomicBool = AtomicBool::new(cfg!(target_os = "linux"));
```
Comment (nagisa) [inline src/libstd/sys/unix/pipe.rs]: Preexisting: 

`pipe2` is also supported on freebsd, netbsd, dragonflybsd, openbsd. These probably should also appear in that `cfg!` above.

**Comment by alexcrichton** [comment]
Looking back on this, I'm not actually sure why handling ENOSYS is here at all. We're not doing a syscall manually, but instead calling a glibc function. That means if it exists we won't get ENOSYS from what I understand. In that case can we just remove that branch?

**Comment by tbu-** [comment]
True.

**Comment by tbu-** [comment]
@alexcrichton Removed the branch.

**Comment by alexcrichton** [comment]
@bors: retry

* sccache wedged #39003

**Comment by tbu-** [comment]
Some `curl` failed, looks unrelated.

**Comment by cuviper** [comment]
> Looking back on this, I'm not actually sure why handling ENOSYS is here at all. We're not doing a syscall manually, but instead calling a glibc function. That means if it exists we won't get ENOSYS from what I understand.

This is not necessarily true.  Glibc has had the `pipe2` symbol since 2.9, but it makes no effort to emulate the functionality if the kernel rejects it.  The [`pipe2` stub](https://sourceware.org/git/?p=glibc.git;a=blob;f=io/pipe2.c;h=ec210ee4a351a031e8de72ff3b06dc30a29dc336;hb=HEAD) does nothing, and the all of the arch-specific implementations are raw syscall wrappers that do nothing but copy returned errors to `errno`.  So if your running kernel happens to be older, you'll just get `ENOSYS`.  (I'll bet you can guess why I'm noticing this change now...)

**Comment by alexcrichton** [comment]
Ack sorry about that @cuviper! The fix here would be to just reinstate `ENOSYS` handling to fall back to `pipe`, right?

**Comment by cuviper** [comment]
Yeah, that should do it!

**Comment by alexcrichton** [comment]
Ok I'll send a PR for this. Looks like this made it into 1.17.0 but we can at least backport and get it fixed for 1.19.0

**Comment by alexcrichton** [comment]
https://github.com/rust-lang/rust/pull/42521

**Pull request merged**